fx quant: make sure observer is inserted before a quantized output

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49420 fx quant: make sure observer is inserted before a quantized output**
* #49382 fx quant: fix fq when input is quantized and node does not need fq
* #49239 fx quant: do not insert observers at quantized inputs
* #49238 fx quant: move {input|output}_quantized_idxs cfg from convert to prepare
* #49213 eager quant: remove fake_quant after add/mul nodes during QAT

Summary:

Before: if an output was marked as quantized, it could actually not
be quantized, if the previous node was not quantized.

After: if an output was marked as quantized, it will be quantized
regardless of the quantization status of the previous node.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_quant_output_always_observed
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25566834](https://our.internmc.facebook.com/intern/diff/D25566834)

vkuzo committed:
fx quant: make sure observer is inserted before a quantized output

Summary:

Before: if an output was marked as quantized, it could actually not
be quantized, if the previous node was not quantized.

After: if an output was marked as quantized, it will be quantized
regardless of the quantization status of the previous node.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_quant_output_always_observed
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fx/quantize.py:
```diff
@@ -399,15 +399,32 @@ def load_arg(a):
+                output_node_seen_cnt += 1
+                if cur_output_node_idx in output_quantized_idxs:
+                    prev_node = node.args[0]
+                    if prev_node.name not in observed_node_names_set:
```
should we assert instance(prev_node, Node) here? output could be dict/tuple etc. I think.

jerryzh168 reviewed (approved):
Had one inline comment.

vkuzo committed:
comments on "fx quant: make sure observer is inserted before a quantized output"


Summary:

Before: if an output was marked as quantized, it could actually not
be quantized, if the previous node was not quantized.

After: if an output was marked as quantized, it will be quantized
regardless of the quantization status of the previous node.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_quant_output_always_observed
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25566834](https://our.internmc.facebook.com/intern/diff/D25566834)

[ghstack-poisoned]

vkuzo committed:
Update on "fx quant: make sure observer is inserted before a quantized output"


Summary:

Before: if an output was marked as quantized, it could actually not
be quantized, if the previous node was not quantized.

After: if an output was marked as quantized, it will be quantized
regardless of the quantization status of the previous node.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_quant_output_always_observed
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25566834](https://our.internmc.facebook.com/intern/diff/D25566834)

[ghstack-poisoned]

vkuzo committed:
rebase on "fx quant: make sure observer is inserted before a quantized output"


Summary:

Before: if an output was marked as quantized, it could actually not
be quantized, if the previous node was not quantized.

After: if an output was marked as quantized, it will be quantized
regardless of the quantization status of the previous node.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_quant_output_always_observed
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25566834](https://our.internmc.facebook.com/intern/diff/D25566834)

[ghstack-poisoned]
